src/Orleans.Core.Abstractions/IDs/GrainId.cs

```diff
@@ -55,15 +55,15 @@ public GrainId(SerializationInfo info, StreamingContext context)
 
         public readonly bool IsDefault => _type.IsDefault && _key.IsDefault;
 
-        public override bool Equals(object obj) => obj is GrainId id && this.Equals(id);
+        public override readonly bool Equals(object obj) => obj is GrainId id && this.Equals(id);
```

Comment:
It allows the compiler to avoid defensive copies in some cases (eg, when the struct is held in a readonly field)

Reply:
I wrote some more about this in a blog post: https://reubenbond.github.io/posts/dotnet-perf-tuning

Under the heading "Avoid defensive copies"

Reply:
Except this is a `readonly struct` already, so shouldn't it always be able to elide those? And indeed, a quick check shows that, adding the readonly to methods on a readonly struct makes no difference to the output IL.

So specifying readonly explicitly here would be mostly for consistency (handle on-writing methods the same in bot mutable and immutable structs), and also helps if any of these functions get copied to a mutable struct, since there we really would want them marked readonly for copy elision reasons.

Reply:
Oh, sorry, yes, in this case readonly is implicit so it will have no effect